File: exam_at_home/1/ugly_number.py
#
# author: wang-yang
# email: tnst92002@gmail.com
#
from heapq import *

ug = [0]
pq = [1]

while len(ug) <= 10000:
    v = heappop(pq)
    if v == ug[-1]:
        continue
    ug.append(v)
    heappush(pq, v*2)
    heappush(pq, v*3)
    heappush(pq, v*5)

N = int(input())
for _ in range(N):
    k = int(input())
    print(ug[k])

#
# author: jamie
# email: jinjiedeng.jjd@gmail.com
#
Q = int(input())
Ks = []
for _ in range(Q):
    Ks.append(int(input()))
max_k = max(k for k in Ks)
ugly_number = [0] * (max_k + 1)
i_2, i_3, i_5 = 1, 1, 1
ugly_number[1] = 1
for i in range(2, max_k + 1):
    x_2, x_3, x_5 = ugly_number[i_2] * 2, ugly_number[i_3] * 3, ugly_number[i_5] * 5
    ugly_number[i] = min(x for x in [x_2, x_3, x_5])
    if ugly_number[i] == x_2: i_2 += 1
    if ugly_number[i] == x_3: i_3 += 1
    if ugly_number[i] == x_5: i_5 += 1
for k in Ks:
    print(ugly_number[k])

File: exam_at_home/1/ugly_number.cc
/*
* author: wang-yang
* email: tnst92002@gmail.com
*/
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    priority_queue<ll, vector<ll>, greater<ll>> pq;
    vector<ll> ug {0};
    pq.emplace(1);
    while (ug.size() <= 10000) {
        ll nxt = pq.top();
        pq.pop();
        if (nxt == ug.back()) continue;
        ug.emplace_back(nxt);
        pq.emplace(nxt * 2);
        pq.emplace(nxt * 3);
        pq.emplace(nxt * 5);
    }
    int N, k;
    cin >> N;
    while (N--) {
        cin >> k;
        cout << ug[k] << '\n';
    }
}

/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <algorithm>
#include <iostream>
#include <vector>

int main ()
{
  uint64_t Q;
  std::cin >> Q;
  std::vector<uint64_t> Ks(Q);
  for (auto& k : Ks) { std::cin >> k; }
  auto max_k {*max_element(Ks.begin(), Ks.end())};
  std::vector<uint64_t> ugly_number(max_k + 1);
  auto it {std::next(ugly_number.begin())};
  auto it_2 {it}, it_3 {it}, it_5 {it};
  *it++ = 1;
  while (it != ugly_number.end())
  {
    auto x_2 {*it_2 * 2}, x_3 {*it_3 * 3}, x_5 {*it_5 * 5};
    auto min {std::min(x_2, std::min(x_3, x_5))};
    *it++ = min;
    if (min == x_2) { ++it_2; }
    if (min == x_3) { ++it_3; }
    if (min == x_5) { ++it_5; }
  }
  for (auto k : Ks) { std::cout << ugly_number[k] << "\n"; }
  return 0;
}